Sparse CSR CUDA: Add block torch.addmm when mat1 is sparse

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68711
* #68710
* #68709
* #68708
* __->__ #68707

This PR adds a path for block CSR matrices for `torch.addmm`. cuSPARSE interface is restricted to 32-bit indices and square blocks.
My plan is to make everything work and tests passing using an unsafe constructor first, keeping it all private. Then discuss & implement constructors with block information separately unlocking the functions for wider use. Documentation will come with the update to constructors.

cc @nikitaved @pearu @cpuhrsch @IvanYashchuk @ngimel

Differential Revision: [D32650366](https://our.internmc.facebook.com/intern/diff/D32650366)

IvanYashchuk committed:
Sparse CSR CUDA: Add block torch.addmm when mat1 is sparse

This PR adds a path for block CSR matrices. cuSPARSE is restricted to
32-bit indices and square blocks.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add block torch.addmm when mat1 is sparse"


This PR adds a path for block CSR matrices for `torch.addmm`. cuSPARSE interface is restricted to 32-bit indices and square blocks.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

cpuhrsch commented on test/test_sparse_csr.py:
```diff
@@ -575,6 +579,50 @@ def test_csr_matvec(self, device, dtype):
+                nnz = random.randint(0, m * k)
+                a = self.genSparseCSRTensor((m, k), nnz, dtype=dtype, device=device, index_dtype=index_dtype)
+                a_data = make_tensor((nnz, block_size, block_size), dtype=dtype, device=device)
+                a = torch._sparse_csr_tensor_unsafe(a.crow_indices(), a.col_indices(), a_data, (m * block_size, k * block_size))
```
To be precise, this means the CSR constructor will have different semantics from the COO constructor, right?

Because for COO, you'll need to introduce dense dimensions to support higher dimensional values, i.e.

```
>>> i = [[0, 0, 1], [0, 1, 1]]
>>> v = torch.randn(3, 2, 2)
>>> a = torch.sparse_coo_tensor(i, v, (2, 2))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: number of dimensions must be sparse_dim (2) + dense_dim (2), but got 2
>>> a = torch.sparse_coo_tensor(i, v, (2, 2, 2, 2))
>>> a.size()
torch.Size([2, 2, 2, 2])
```

I also understand that you're using the private, unsafe constructor. Does this mean you want to change the behavior or arguments passed to the public constructor later on?

In general I'm in favor these semantics. What we want with block sparse layouts is tiling and not additional dense dimensions in the [Hybrid COO sense](https://pytorch.org/docs/master/sparse.html#hybrid-sparse-coo-tensors).

The only issue is that we don't have metadata right now that distinguishes between dimensions of the data used for tiling vs. dimensions exposed to the user as dense dimensions. [The number of dense dimensions is inferred from the dimensionality of the values](https://github.com/pytorch/pytorch/blob/2f099c7555ab41e55ddfd12ba3ed804e52894820/aten/src/ATen/native/sparse/SparseTensor.cpp#L281-L290). We will want dense dimensions soon as well to help represent data points such as embeddings of sentences of different lengths. 

So, for example, let's say you have 4 Tensors of size ```(3, 128), (15, 128), (6, 128), (3, 128)```. Here a CSR Tensor of shape ```(4, 15, 128)```, where the last dimension is dense, is very useful.

IvanYashchuk replied:
Yes, the current use of an unsafe CSR constructor is different from the COO. Of course, the public constructor of CSR matrices and conversion functions will need to be modified to be able to create inputs for these new paths.
My plan is to make everything work and tests passing using an unsafe constructor first, keeping it all private. Then discuss & implement constructors with block information separately unlocking the functions for wider use. Documentation will come with the update to constructors.

 > We will want dense dimensions soon as well to help represent data points such as embeddings of sentences of different lengths.
So, for example, let's say you have 4 Tensors of size (3, 128), (15, 128), (6, 128), (3, 128). Here a CSR Tensor of shape (4, 15, 128), where the last dimension is dense, is very useful.

Please put this information into a separate feature request issue.
I know that the most generic representation, that is possible to create in PyTorch, is a CSR matrix of shape `(rows, columns, *dense_shape)`, where rows and columns represent matrix dimensions and `dense_shape` can be shape of any length. However, tensors with this generic representation wouldn't be usable with math libraries like MKL Sparse, cuSPARSE. The libraries expect the `dense_shape` to be strictly either `()` for standard CSR format or `(block_size, block_size)` for standard BSR format.

IvanYashchuk commented:
ROCm failure is real. I should update the code to make it work with the hipSPARSE library.

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add block torch.addmm when mat1 is sparse"


This PR adds a path for block CSR matrices for `torch.addmm`. cuSPARSE interface is restricted to 32-bit indices and square blocks.
My plan is to make everything work and tests passing using an unsafe constructor first, keeping it all private. Then discuss & implement constructors with block information separately unlocking the functions for wider use. Documentation will come with the update to constructors.

cc nikitaved pearu cpuhrsch IvanYashchuk ngimel

[ghstack-poisoned]

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32633806).

cpuhrsch commented on aten/src/ATen/native/LinearAlgebraUtils.h:
```diff
@@ -30,6 +30,14 @@ static char to_blas(TransposeType trans) {
   TORCH_INTERNAL_ASSERT(false, "Invalid transpose type");
 }
 
+static inline c10::MaybeOwned<Tensor> expect_resolved_conj(const Tensor& tensor) {
```
Why is this function declared as static?

malfet commented:
Hmm, looks like newly added tests clearly breaking CI, see https://github.com/pytorch/pytorch/runs/4308836975?check_suite_focus=true

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32650366).

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add block torch.addmm when mat1 is sparse"


This PR adds a path for block CSR matrices for `torch.addmm`. cuSPARSE interface is restricted to 32-bit indices and square blocks.
My plan is to make everything work and tests passing using an unsafe constructor first, keeping it all private. Then discuss & implement constructors with block information separately unlocking the functions for wider use. Documentation will come with the update to constructors.

cc nikitaved pearu cpuhrsch IvanYashchuk ngimel

Differential Revision: [D32650366](https://our.internmc.facebook.com/intern/diff/D32650366)

[ghstack-poisoned]

IvanYashchuk commented:
@malfet, @cpuhrsch I added a fix and verified locally that tests now pass on Linux CUDA 10.2.

IvanYashchuk commented on aten/src/ATen/cuda/CUDASparseDescriptors.cpp:
```diff
@@ -71,7 +71,7 @@ CuSparseDnMatDescriptor::CuSparseDnMatDescriptor(const Tensor& input) {
   auto leading_dimension =
       is_row_major ? input_strides[ndim - 2] : input_strides[ndim - 1];
-  auto order = is_row_major ? CUSPARSE_ORDER_ROW : CUSPARSE_ORDER_COL;
+  auto order = is_column_major ? CUSPARSE_ORDER_COL : CUSPARSE_ORDER_ROW;
```
This change fixes CUDA 10.2 failure.

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32650366).

malfet commented:
Looks like it causes bunch of failures with CUDA-11.1 now

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add block torch.addmm when mat1 is sparse"


This PR adds a path for block CSR matrices for `torch.addmm`. cuSPARSE interface is restricted to 32-bit indices and square blocks.
My plan is to make everything work and tests passing using an unsafe constructor first, keeping it all private. Then discuss & implement constructors with block information separately unlocking the functions for wider use. Documentation will come with the update to constructors.

cc nikitaved pearu cpuhrsch IvanYashchuk ngimel

Differential Revision: [D32650366](https://our.internmc.facebook.com/intern/diff/D32650366)

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add block torch.addmm when mat1 is sparse"


This PR adds a path for block CSR matrices for `torch.addmm`. cuSPARSE interface is restricted to 32-bit indices and square blocks.
My plan is to make everything work and tests passing using an unsafe constructor first, keeping it all private. Then discuss & implement constructors with block information separately unlocking the functions for wider use. Documentation will come with the update to constructors.

cc nikitaved pearu cpuhrsch IvanYashchuk ngimel

Differential Revision: [D32650366](https://our.internmc.facebook.com/intern/diff/D32650366)

[ghstack-poisoned]

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32650366).
